### PR TITLE
Refactoring. V2.2.0 Circom Support

### PR DESCRIPTION
**Pull request by KyrylR**

### CODE REVIEW

File: grammar/CircomParser.g4 [hunk 1]
```diff
@@ -3,171 +3,219 @@ parser grammar CircomParser;
...
+    : 'component' 'main' publicInputsDefinition? '=' ID '(' argValues=expressionList? ')' ';'
+    ;
 
-forUpdate: ID (SELF_OP | ((ASSIGNMENT | ASSIGNMENT_WITH_OP) expression)) | SELF_OP ID ;
+/*//////////////////////////////////////////////////////////////
+                           STATEMENTS
+//////////////////////////////////////////////////////////////*/
 
-parExpression: '(' expression ')' ;
+body: '{' statments* '}';
```
Comment (Arvolear) [inline grammar/CircomParser.g4]: Is this deliberate?

```suggestion
body: '{' statements* '}';
```

File: grammar/CircomParser.g4 [hunk 2]
```diff
@@ -3,171 +3,219 @@ parser grammar CircomParser;
...
+ifStatments
+    : 'if' '(' cond=expression ')' ifStatments                                  #IfWithFollowUpIf
+    | 'if' '(' cond=expression ')' regularStatmetns                             #IfRegular
+    | 'if' '(' cond=expression ')' regularStatmetns 'else' ifStatments          #IfRegularElseWithFollowUpIf
+    | 'if' '(' cond=expression ')' regularStatmetns 'else' regularStatmetns     #IfRegularElseRegular
     ;
 
-logStmt
-    : 'log' '(' ((STRING | expression) (',' (STRING | expression))*)? ')'
+regularStatmetns
```
Comment (Arvolear) [inline grammar/CircomParser.g4]:
```suggestion
regularStatements
```

File: grammar/CircomParser.g4 [hunk 3]
```diff
@@ -3,171 +3,219 @@ parser grammar CircomParser;
...
+    | '[' expressionList ']'                                                            #PArray
+    | ID '(' expressionList? ')'                                                        #PCall
+    | ID '(' expressionList? ')' '(' (expressionList | expressionListWithNames)? ')'    #PAnonymousCall
     ;
 
-rhsValue
-    : '(' expressionList ')'
-    | expression
-    | blockInstantiation
+assignmentExpression
```
Comment (Arvolear) [inline grammar/CircomParser.g4:176]: Are `==>` and `-->` missing?

File: grammar/CircomParser.g4 [hunk 4]
```diff
@@ -3,171 +3,219 @@ parser grammar CircomParser;
...
-    | '(' expression RIGHT_CONSTRAINT ID (',' expression RIGHT_CONSTRAINT ID)* ')'
+/*//////////////////////////////////////////////////////////////
+                           IDENTIFIER
+//////////////////////////////////////////////////////////////*/
+
+varIdentifier: identifier ('=' rhs=expression)? ;
+
+varIdentifierList: (varIdentifier ',')* varIdentifier ;
+
+signalIdentifier
```
Comment (Arvolear) [inline grammar/CircomParser.g4:190]: Same here

File: grammar/CircomParser.g4 (hunk 3 above)
Comment (KyrylR) [reply grammar/CircomParser.g4:176]: Handled in `substitutions` rule

File: grammar/CircomParser.g4 (hunk 4 above)
Comment (KyrylR) [reply grammar/CircomParser.g4:190]: Handled in `substitutions` rule

**Review by Arvolear — APPROVED** [review]
Looks promising

**Pull request merged**